identity_did/src/document/traits.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::Serialize;
+
+use identity_core::crypto::GetSignature;
+
+use crate::did::DID;
+use crate::service::Service;
+use crate::utils::DIDUrlQuery;
+use crate::verifiable::VerifierOptions;
+use crate::verification::MethodScope;
+use crate::verification::VerificationMethod;
+use crate::Result;
+
+// TODO: add sign_data, split sign/verify to separate trait as first step towards
+//       supporting custom signature schemes. Replace DocumentSigner with trait?
+//       Add DocumentMut for mutable function returns?
+//       Blanket impl for &T, &mut T, Box<T> etc.?
+/// Common operations for DID Documents.
+pub trait Document {
+  type D: DID;
+  type U;
+  type V;
+
+  /// Returns a reference to the `Document` id.
+  fn id(&self) -> &Self::D;
+
+  /// Returns the first [`Service`] with an `id` property matching the provided `query`, if present.
+  fn resolve_service<'query, 'me, Q>(&'me self, query: Q) -> Option<&Service<Self::D, Self::V>>
```

Comment:
Yes, it was the previous design direction and makes it convenient to add extra fields to `Service<D, U>` or `VerificationMethod<D, U>` by injecting an arbitrary struct for `U`, and convert between them cheaply without duplicating field definitions. 
Note that even if we eliminate `U` we still rely on the `D: DID` generic parameter everywhere, though we might want to reconsider it and go with `CoreDID` to enable storing identifiers from other methods in our structs for whatever reason. I still don't really comprehend why the fully-qualified DID Url is required for the `id` fields instead of just a fragment in the specification.

The tagged enum approach still has its drawbacks, I would actually consider a `Service` trait to be a potentially better avenue now.

Specifically regarding open question 6: I would be fine removing `T` from `Credential<T>` and `Presentation<T>`, but we  need to come up with a canonical way for extending credentials and presentation contents easily.

Overall this is just something to consider during the bigger refactors, will leave the generics as-is for now.